OpenRA.Mods.Common/Traits/Cloak.cs

```diff
@@ -77,10 +80,30 @@ public void Created(Actor self)
 
 		protected override void UpgradeDisabled(Actor self)
 		{
-			Uncloak();
+			if (IsTraitDisabled != wasDisabled)
+			{
+				Game.Sound.Play(Info.UncloakSound, self.CenterPosition);
+				if (upgradeManager != null)
+					foreach (var u in Info.WhileCloakedUpgrades)
+						upgradeManager.RevokeUpgrade(self, u, this);
+				wasDisabled = IsTraitDisabled;
+			}
+
 			remainingTime = Info.InitialDelay;
 		}
 
+		protected override void UpgradeEnabled(Actor self)
+		{
+			if (IsTraitDisabled != wasDisabled)
+			{
+				Game.Sound.Play(Info.CloakSound, self.CenterPosition);
+				if (upgradeManager != null)
+					foreach (var u in Info.WhileCloakedUpgrades)
+						upgradeManager.RevokeUpgrade(self, u, this);
```

Comment:
Actually... i'm not sure this should be doing anything, since the actual cloaking should be happening later.